File: code/word-count.py
#!/usr/bin/env python3
# Read from stdin and produce a sorted word frequency table

import fileinput
import re

word_counts = {}

# process text from stdin or given file
for line in fileinput.input():
    # conver alphabet characters to lowercase
    # and non-alphabet characters to space,
    # then split into words by space
    words = re.sub(r'[^A-Za-z]', ' ', line.lower()).split()
    # count unique words
    for word in words:
        if word in word_counts:
            word_counts[word] += 1
        else:
            word_counts[word] = 1

# sort words by descending frequency and ascending lexicographical order
words_sorted = sorted(word_counts, key = lambda x: (-word_counts[x], x))

# print word frequency table
for word in words_sorted:
    print('   {} {}'.format(word_counts[word], word))

File: code/word-count.pl
#!/usr/bin/env perl
# Read from stdin and produce a sorted word frequency table

use strict;
use warnings;
use autodie;


my %word_counts;

# read text from stdin or given file
while (<>) {
	# remove trailing new line, operating on $_
	chomp;
	# convert non-alphabetic characters to space, squeezing duplicates, on $_
	tr/A-Za-z/ /cs;
	# convert to lowercase, assign to $_
	$_ = lc;
	# split $_ by space and store results in array
	my @words = split;
	# count unique words
	for my $word (@words) {
		$word_counts{$word}++;
	}
}

# sort words by descending word frequency and ascending lexicographical order
my @words_sorted = sort
	{ $word_counts{$b} <=> $word_counts{$a} || $a cmp $b }
	keys %word_counts;

# print word frequency table
for my $word (@words_sorted) {
	printf "   %s %s\n", $word_counts{$word}, $word
}

